src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,15 +300,44 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app max-command-failure? max-command-size]
+  (fn [req]
+    (if (= :post (:request-method req))
+      (let [length-in-bytes (request/content-length req)]
+
+        (if length-in-bytes
+          (log/infof "Processing command with a content-length of %s bytes" length-in-bytes )
+          (log/warn (str "No content length found for POST. "
+                         "POST bodies that are too large could cause memory-related failures.")))
+
+        (if (and length-in-bytes
+                 max-command-failure?
+                 (> length-in-bytes max-command-size))
+          (do
+            (log/warnf "content-length of command is %sMB and is larger than the maximum allowed command size of %sMB"
```

Comment:
Yeah, I guess thats one way to think about it.

BTW, the digits get crazy when dividing by base2 MB's:

```
2015-12-08 22:04:42,323 INFO  [qtp1834376846-31 - /pdb/cmd/v1?checksum=eeabbc9b27a3995e34d580d00913fc34dc3015c2] [p.p.middleware] Processing command with a content-length of 6180831 bytes
2015-12-08 22:04:42,324 WARN  [qtp1834376846-31 - /pdb/cmd/v1?checksum=eeabbc9b27a3995e34d580d00913fc34dc3015c2] [p.p.middleware] content-length of command is 5.894499778747559MB and is larger than the maximum allowed command size of 2.168292045593262MB
```

Hilarious :-). I think a rounding of 2-3 decimal places would totally work here in this case.